ARROW-13986: [Go][Parquet] Add File Writers and tests

zeroshade opened the pull request.

emkornfield commented on go/parquet/encryption_read_config_test.go:
```diff
@@ -0,0 +1,443 @@
+	// if we get decryption_config_num = x then it means the actual number is x+1
+	// and since we want decryption_config_num=4 we set the condition to 3
+	props := parquet.NewReaderProperties(memory.DefaultAllocator)
+	if decryptConfigNum != 3 {
```
it might be more cleaner to encapsulate the test case into its own class/struct instead of the if/then here.  At the end of the day that might be just a slightly different form of obfuscation.

emkornfield commented on go/parquet/encryption_read_config_test.go:
```diff
@@ -0,0 +1,443 @@
+		}
+		d.EqualValues(i, int96md.NumValues())
+
+		if decryptConfigNum != 3 {
```
i think it is worth commenting here and above why this logic doesn't apply to case number 3

zeroshade replied:
added a comment

emkornfield commented on go/parquet/encryption_read_config_test.go:
```diff
@@ -0,0 +1,443 @@
+			expected[8] = byte('0') + byte(i/10)%10
+			expected[9] = byte('0') + byte(i%10)
+			if i%2 == 0 {
+				d.Equal(1, valuesRead)
```
what is this verifying exactly?  Should this be augmented with a test verifying errors/inequality when reading data from an encrypted file without keys?

zeroshade replied:
The lines here: https://github.com/apache/arrow/pull/11538/files/61ce47474d7ec42a80f1d6b354fbc28f4914edab#diff-cef277f99a3822e591750fc26b274b33e7be004a21833b2f38f5c438c1069178R382

and here: https://github.com/apache/arrow/pull/11538/files/61ce47474d7ec42a80f1d6b354fbc28f4914edab#diff-cef277f99a3822e591750fc26b274b33e7be004a21833b2f38f5c438c1069178R391

Verify that a panic happens when attempting to read from an encrypted file without proper keys.

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+
+	// the total number of stored values. for repeated or optional values. this
+	// number may be lower than numBuffered
+	numBufferedEncoded int64
```
numDataValues might be a better name?

zeroshade replied:
done

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	// non-repeated, required columns, this is equal to the number of encoded
+	// values. For repeated or optional values, there may be fewer data values
+	// than levels, and this tells you how many encoded levels there are in that case
+	numBuffered int64
```
numBufferedValues or numBufferedLevels might be better?

zeroshade replied:
done

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	// records the current number of compressed bytes in a column
+	totalCompressedBytes int64
+	closed               bool
+	fallback             bool
```
is this fallback to non dictionary?  could the name be improved?

zeroshade replied:
renamed

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	}
+
+	if ret.props.DataPageVersion() == parquet.DataPageV1 {
+		if ret.descr.MaxDefinitionLevel() > 0 {
```
maybe add a comment on why the offset is being added (this is so the number of levels can be recorded at the beginning of the buffer?)

zeroshade replied:
comment added

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	return w.totalBytesWritten
+}
+
+func (w *columnWriter) RowsWritten() int64 {
```
is there a strong reason to up-cast this to int64?

zeroshade replied:
I was being consistent with the other functions (returning numbers of bytes) but to be fair, it kinda makes sense to just leave this as an `int`.

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	w.defLevelSink.Reset(0)
+	w.repLevelSink.Reset(0)
+
+	if w.props.DataPageVersion() == parquet.DataPageV1 {
```
this seems duplicative with the code above?  Can they be consolidated?

zeroshade replied:
consolidated by having the above code call the `reset` function (formerly `init`)

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	w.repEncoder.EncodeNoFlush(levels)
+}
+
+func (w *columnWriter) init() {
```
would reset be a better name for this function?

zeroshade replied:
renamed

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	w.repEncoder.Reset(w.descr.MaxRepetitionLevel())
+}
+
+func (w *columnWriter) concatBuffers(defLevelsSize, repLevelsSize int64, values []byte, wr io.Writer) {
```
nit: would flushBuffers be a better name for this method?

emkornfield replied:
shouldn't repLevels and defLevels always have the same size?

zeroshade replied:
These is the encoded buffers which means that they could potentially be different sizes due to the run-length encoding since the values aren't necessarily the same resulting in differently sized buffers at the end.

Also this isn't really flushing them, it's creating a single contiguous buffer that would then get flushed or compressed and then flushed. It's just concatenating the three buffers (encoded def levels, encoded rep levels and raw values) into a single buffer.

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	}
+	defer values.Release()
+
+	isV1DataPage := w.props.DataPageVersion() == parquet.DataPageV1
```
maybe make this a helper method in general?

zeroshade replied:
you mean add a `IsV1DataPage()` function to the properties struct?

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	return nil
+}
+
+func (w *columnWriter) AddDataPage() error {
```
Would FlushCurrentPage be a better name for this method?

zeroshade replied:
renamed

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	}
+
+	// concatenate uncompressed levels and the possibly compressed values
+	combined := make([]byte, int(defLevelsRLESize+repLevelsRLESize+int64(len(data))))
```
is there a reason not to use concatBuffers here?

zeroshade replied:
modified to use `concatBuffers`

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+	return toWrite
+}
+
+func (w *columnWriter) writeLevelsSpaced(numLevels int64, defLevels, repLevels []int16) {
```
is there a reason to not re-use the logic above?

emkornfield replied:
i guess it avoid  calculated valuestowrite?

zeroshade replied:
yea, that's the only reason it's not reusing the logic above it. no need to calculate values to write.

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+// outValsToWrite, outSpacedValsToWrite, and NullCount. Additionally
+// it will update the validity bitmap if required (i.e. if at least one
+// level of nullable structs directly preced the leaf node)
+func (w *columnWriter) maybeCalculateValidityBits(defLevels []int16, batchSize int64, valuesToWrite, spacedValuesToWrite, nullCount *int64) {
```
instead of taking the output values as args could we define a struct for them and return that?

zeroshade replied:
done

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+				}
+			}
+			*valuesToWrite += toWrite
+			*spacedValuesToWrite += spacedToWrite
```
is spacedValuesToWrite needed would just returning valuesToWrite and nullcount be sufficient?

zeroshade replied:
technically you're correct (and i've modified the code as such) having them as three separate values did enable a couple fewer subtractions/additions but in theory the compiler should be able to recognize and optimize. I haven't benchmarked it yet but I wouldn't expect this to be an issue. If future benchmarking shows it as a potential performance improvement then I'll modify it at that point. for now the simplified code is better.

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,567 @@
+
+	if values.Data().Offset() > 0 {
+		data := values.Data()
+		buffers[1] = memory.NewBufferBytes(data.Buffers()[1].Bytes()[data.Offset()*arrow.Int32SizeBytes : data.Len()*arrow.Int32SizeBytes])
```
this might be based on old buggy code, please take a look at: https://github.com/apache/arrow/blob/040bd1e8d563bed532b712721078757f4aa0907e/cpp/src/parquet/column_writer.cc#L1321

I think the actual validity buffer needs to be sliced as well.

zeroshade replied:
looking at the file you linked, it looks like what i was missing was that i needed to do `buffers[0] = w.bitsBuffer` which is what the code linked does. Which makes sense as the `w.bitsBuffer` should already be updated with the correct slice of the validity buffer we want.

emkornfield replied:
yeah, please look at the blame and look at the additional unit tests added to make sure it is covered.

emkornfield reviewed (changes requested):
Sorry still going through this but left some initial comments.

zeroshade commented:
@emkornfield  thanks for going through this, i've updated as per your comments so far.

zeroshade commented:
rebased this. @emkornfield lemme know if there's any other comments you have for this. thanks!

emkornfield commented:
sorry will try to carve out some time this weekend to finish reviewing

zeroshade commented:
bump @emkornfield

emkornfield commented:
Sorry things have been more hectic then I would have liked recently, will try to get to this soon.  Thank you for your patience.

zeroshade commented:
@emkornfield Sorry I've been out for a few weeks + a week for being sick. I'd love to be able to get this merged in time for the v7 release if at all possible. Would you have any time soon to be able to finish your review?

emkornfield commented on go/parquet/file/page_writer.go:
```diff
@@ -0,0 +1,466 @@
+	hdr := dataPageV2HeaderPool.Get().(*format.DataPageHeaderV2)
+	hdr.NumValues = page.nvals
+	hdr.NumNulls = page.nulls
+	hdr.NumRows = page.nrows
```
Can you check if https://issues.apache.org/jira/browse/PARQUET-2066 applies here?

emkornfield replied:
I guess it would be in the caller.

zeroshade replied:
Good catch, you're right. The `buildDataPageV2` function passes `numValues` for both `nvals` and `nrows`. I'll see about fixing that so it handles nested values properly to get the correct number of rows.

zeroshade replied:
Commit 2686ea26fb93c676825c3ba54e55b895202784c5 fixes the issue with num rows and adds a unit test for it.

emkornfield commented on go/parquet/file/page_writer.go:
```diff
@@ -0,0 +1,466 @@
+
+	pageHdr := pageHeaderPool.Get().(*format.PageHeader)
+	defer pageHeaderPool.Put(pageHdr)
+	pageHdr.UncompressedPageSize = int32(uncompressed)
```
is overflow possible here?

zeroshade replied:
so i just double checked the `parquet.thrift` file and realized that the `uncompressed_size` for data pages is an int32, not an int64. So i've updated the `DataPage` structs accordingly to be int32 for the sizes of the pages the whole way through, making sure that we avoid overflows since we'll always read an int32 from the file and the writer should enforce the max page size in the options.

emkornfield commented on go/parquet/internal/testutils/primitive_typed.go:
```diff
@@ -0,0 +1,305 @@
+		}
+		return n1
+	}
+	panic("wtf")
```
is wtf common in Go panics?

zeroshade replied:
🤦  Nope, that's just something I forgot to fix with a more descriptive error message from when i was building this. Haha, i'll fix that.

zeroshade replied:
fixed

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,570 @@
+
+func (b batchWriteInfo) numSpaced() int64 { return b.batchNum + b.nullCount }
+
+// this will always update the three otuput params
```
typo:
```suggestion
// this will always update the three output params
```

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -0,0 +1,570 @@
+// this will always update the three otuput params
+// outValsToWrite, outSpacedValsToWrite, and NullCount. Additionally
+// it will update the validity bitmap if required (i.e. if at least one
+// level of nullable structs directly preceed the leaf node)
```
```suggestion
// level of nullable structs directly precede the leaf node)
```

emkornfield commented:
A lot of code so I might have missed something but I don't think I have any more comments.  It would be really great to figure out an integration testing story with other implementations to help verify correctnes.

zeroshade commented:
@emkornfield I'll update based on the comments, but as far as integration testing, currently the integration testing I have is that I use several of the test parquet files from here: https://github.com/apache/parquet-testing/tree/acd375eb86a81cd856476fca0f52ba6036a067ff/data in unit tests by setting PARQUET_TEST_DATA just like the C++ implementation. I don't think I implemented tests for *all* of the files there, but a large amount of them are currently tested, and I'll probably add more as I go.

But I definitely agree that a more robust cross-implementation testing suite would be a great addition.

zeroshade commented:
@emkornfield Updated from feedback, rebased/merged changes from master. Let me know if you're happy with the changes/fixes I made from your comments. Thanks again!

emkornfield commented:
Also I forget if I CCed you on https://issues.apache.org/jira/browse/PARQUET-2067

zeroshade commented:
You didn't, but the test I added to solve the issue addresses that too, it verifies both that there's 2 nulls reported and 3 rows.

zeroshade commented:
@emkornfield just bumping to make sure you see I addressed the comments you left. If i don't hear back from you in a few days i'll merge this as you said you don't think you have any more comments, hopefully you'll have some time to give the latest changes a quick look and mark this approved or not. 

Thanks again for dealing with this huge undertaking of reviewing this. There's only one package left after this gets merged and then it's all in!

emkornfield commented on go/parquet/file/column_writer.go:
```diff
@@ -179,7 +179,7 @@ func (w *columnWriter) TotalBytesWritten() int64 {
 
 func (w *columnWriter) RowsWritten() int {
-	return w.rowsWritten
+	return w.rowsWritten + w.numBufferedRows
```
not sure if this is needs a doc update or not?

zeroshade replied:
the documentation is still correct, I had to add that because of the way I split it up to ensure proper writing of rows vs values in the data pages themselves. `RowsWritten` is still the number of rows written to the column (but not necessarily flushed yet).

emkornfield commented on go/parquet/file/column_writer_test.go:
```diff
@@ -45,6 +49,73 @@ const (
+	return m.Called().Error(0)
+}
+
+func TestWriteDataPageV2NumRows(t *testing.T) {
```
this seems OK, one other note of something broken in C++ for V2 is rows are supposed to begin and end on page boundaries (V1 doesn't have this gaurantee)

zeroshade replied:
that's an interesting one that makes the logic for `WriteBatch` a bit more complex and annoying. I guess that's probably why it's broken for C++ currently haha. I'll see if i can figure out a way to slightly modify the logic for the batching so that I always do batches on row boundaries for V2. if not then i'll just document the limitation, if it's currently broken in the C++ too then I don't feel bad about kicking that can down the road for now in the interest of getting this merged. I'll comment back with what I end up doing.

zeroshade replied:
Ok cool, I solved it by mucking with the batching logic to ensure that the batches in `WriteBatch` always begin at a row boundary (writing slightly smaller batches if necessary). this lets everything else stay the same since as long as I guarantee that the batch I write starts with a new row, and we only check to write the page after writing a batch, it guarantees that we never have a row crossing page boundaries.

the only drawback is that doing it this way means that if you are writing a repeated column with DataPageV2, `WriteBatch` must always start with a new row. that keeps the guarantee. I'm ok with that though and I've updated the documentation accordingly and added a test to ensure that batching makes sure that rows don't cross page boundaries.

Assuming nothing goes wrong with the tests in the CI, i'll merge this after everything passes.

emkornfield commented on go/parquet/file/column_writer_test.go:
```diff
@@ -45,6 +49,73 @@ const (
+		}
+
+		// only match if the page being written has 2 nulls, 6 values and 3 rows
+		return !pagev2.IsCompressed() &&
```
it would be nice to have a similar test for V1 with regards to null count.

zeroshade replied:
The header for V1 data pages doesn't have a null count in it though, so i'm not sure what a similar test would look like other than testing the number of values

emkornfield replied:
I was thinking about the null_count in statistics.

zeroshade replied:
Ah, Completely forgot about that! good point. I've added the stats check for the null count to the tests and updated the test. Thanks!

emkornfield reviewed (approved):
thanks for making the changes a few more notes/potential additional test coverage.
